From 78d6012e8e0892a25784665466688ad3e4004653 Mon Sep 17 00:00:00 2001
From: Paul Naoumenko <paulx.naoumenko@intel.com>
Date: Fri, 5 Oct 2012 18:16:56 +0200
Subject: [PORT FROM MAIN] Wifi: Add robustness to calibration phase

BZ: 59713

Modified calibrator code to return an error code when calibration fails
and when the nvs file is not generated.

Modified wlan_prov in order to try to rexecute calibration
3 times  if failures occur. If after 3 times calibration still fails,
reboot the board.

Change-Id: Iacd00a2b684f36051726475a42db4500c4a616ec
Orig-Change-Id: I9824398503dabc89fbed221f1d87d8bd2745d242
Signed-off-by: Paul Naoumenko <paulx.naoumenko@intel.com>
Reviewed-on: http://android.intel.com:8080/69176
Reviewed-by: Regairaz, Julie <julie.regairaz@intel.com>
Reviewed-by: De Cesco, Jonathan <jonathan.de.cesco@intel.com>
Reviewed-by: Trivelly, Jean <jean.trivelly@intel.com>
Tested-by: Champciaux, Nicolas <nicolas.champciaux@intel.com>
Reviewed-by: buildbot <buildbot@intel.com>
Tested-by: buildbot <buildbot@intel.com>
Reviewed-on: http://android.intel.com:8080/77991
Reviewed-by: Li, XiaojingX <xiaojingx.li@intel.com>
Tested-by: Li, XiaojingX <xiaojingx.li@intel.com>
---
 calibrator/plt.c              |    2 +-
 wlan_prov/wlan_provisioning.c |   23 +++++++++++++++++++----
 2 files changed, 20 insertions(+), 5 deletions(-)

diff --git a/calibrator/plt.c b/calibrator/plt.c
index 27ff9ce..7c14b16 100644
--- a/calibrator/plt.c
+++ b/calibrator/plt.c
@@ -1296,7 +1296,7 @@ out_rmmod:
 out_removenvs:
 	fprintf(stderr, "Calibration not complete. Removing half-baked nvs\n");
 	unlink(cmn.nvs_name);
-	return 0;
+	return res;
 
 }
 COMMAND(plt, autocalibrate, "<dev> <module path> <ini file1> <nvs file> "
diff --git a/wlan_prov/wlan_provisioning.c b/wlan_prov/wlan_provisioning.c
index 9531497..c5a153d 100644
--- a/wlan_prov/wlan_provisioning.c
+++ b/wlan_prov/wlan_provisioning.c
@@ -44,6 +44,7 @@ const char WLAN_SDIO_BUS_PATH[] = "/sys/bus/sdio/drivers/wl1271_sdio/";
 #define SYSFS_SDIO_DEVICES_PATH "/sys/bus/sdio/devices/"
 #define NEW_NVS_FILE_NAME		WIFI_PATH"/new-nvs.bin"
 #define TQS_FILE				"/etc/wifi/TQS.ini"
+#define MAX_CALIBRATION_TRIES	3
 
 /* pattern MAC address in NVS file */
 #define NVS_LENGTH_TO_SET       0x01
@@ -267,11 +268,25 @@ int main(int argc, char **argv)
 	nvsBinFile = fopen(NVS_file_name, "rb");
 
 	if (!nvsBinFile) {
-		LOGI("run calibration");
+		int nbCalibrationTries = 1;
+		LOGI("running calibration, try: %d",nbCalibrationTries);
 		unbind_bind_request = 1;
-		if (wifi_calibration()) {
-			res =  -2;
-			goto end;
+		while (wifi_calibration()) {
+			nbCalibrationTries++;
+			if(nbCalibrationTries >= MAX_CALIBRATION_TRIES) {
+				LOGI("Rebooting after %d calibration tries", MAX_CALIBRATION_TRIES);
+				goto fatal; //Reboot after 3 failed calibrations.
+			}
+			if (unbind_wlan_sdio_drv(WLAN_SDIO_BUS_PATH, device_id)
+						|| bind_wlan_sdio_drv(WLAN_SDIO_BUS_PATH, device_id))
+					{
+						/*
+						*Rebooting:  calibration failed, unbind/bind failed
+						*/
+						LOGI("Rebooting: unbind/bind failed");
+						goto fatal;
+					}
+			LOGI("running calibration, try: %d",nbCalibrationTries);
 		}
 	} else {
 		fclose(nvsBinFile);
-- 
1.7.4.1

